### PR TITLE
Implement crypto for linux os-spec mod

### PR DESCRIPTION
**Pull request by cqs21**
Early discussions: #21

### CODE REVIEW

**Comment by tatetian** [comment]
@lucassong-mh Could you review this PR first? Thanks.

**Comment by lucassong-mh** [comment]
> @lucassong-mh Could you review this PR first? Thanks.

@tatetian No problem. Will go through the code again. My first round of review is in https://github.com/asterinas/mlsdisk/pull/21.

File: linux/bindings/src/crypto/aead.rs [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+    pub fn encrypt(
+        self,
+        aad: &[u8],
+        input: &[u8],
+        iv: &[u8],
+        output: &mut [u8],
+        mac: &mut [u8],
+    ) -> Result {
+        let mut src = SgTable::alloc(Self::SCATTERLIST_SIZE)?;
+        let mut dst = SgTable::alloc(Self::SCATTERLIST_SIZE)?;
```
Comment (lucassong-mh) [inline linux/bindings/src/crypto/aead.rs:267]: Consider the error handling part here: If `dst` failed in allocation while `src` succeeded, would the corresponding SgTable's memory of `src` be freed properly?

File: linux/bindings/src/crypto/skcipher.rs [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+    pub fn alloc_request(&self) -> Result<SkcipherReq> {
+        // SAFETY: `self.skcipher` is a valid cipher.
+        let req = unsafe { crate::skcipher_request_alloc(self.skcipher.as_ptr(), GFP_KERNEL) };
+        if req.is_null() {
+            return Err(ENOMEM);
+        }
+
+        Ok(SkcipherReq {
+            // SAFETY: `req` is checked above, it is valid.
+            req: unsafe { NonNull::new_unchecked(req) },
```
Comment (lucassong-mh) [inline linux/bindings/src/crypto/skcipher.rs:121]: If `req` is a valid pointer, is `NonNull::new(req).expect("skcipher req ptr is null!");` a better choice? (Avoid an unsafe block and safety comment)

File: linux/bindings/src/crypto/skcipher.rs [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+        unsafe {
+            let crypto_alg = (*self.skcipher.as_ptr()).base.__crt_alg;
+            let skcipher_alg = crate::container_of!(crypto_alg, crate::skcipher_alg, base);
+            (*skcipher_alg).ivsize as _
+        }
+    }
+
+    /// Returns the size (in bytes) of the request data structure. The `skcipher_request`
+    /// data structure contains all pointers to data required for the skcipher operation.
+    pub fn reqsize(&self) -> usize {
```
Comment (lucassong-mh) [inline linux/bindings/src/crypto/skcipher.rs:138]: Why the sizes here not Rust code style? (no underline)

File: linux/bindings/src/crypto/skcipher.rs [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+        }
+    }
+
+    /// Handles encryption request.
+    ///
+    /// Returns `Ok(())` if the cipher operation was successful, or an error
+    /// code in [`Error`].
+    pub fn encrypt(self, input: &[u8], iv: &[u8], output: &mut [u8]) -> Result {
+        let mut src = SgTable::alloc(Self::SCATTERLIST_SIZE)?;
+        let mut dst = SgTable::alloc(Self::SCATTERLIST_SIZE)?;
```
Comment (lucassong-mh) [inline linux/bindings/src/crypto/skcipher.rs:226]: Same error handling issue as in `Aead`.

File: linux/dm-sworndisk/src/bio.rs [hunk 5]
```diff
@@ -0,0 +1,350 @@
+// SPDX-License-Identifier: GPL-2.0
```
Comment (lucassong-mh) [inline linux/dm-sworndisk/src/bio.rs:1]: Why this PR described with "crypto" contains block device related code? Can they be divided into independent PR?

File: linux/dm-sworndisk/src/bio.rs [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+            // non-null and valid.
+            unsafe {
+                let page = bindings::virt_to_page(bio_vec.virt_addr as _);
+                bindings::__bio_add_page(bio_ptr, page, bio_vec.len as _, bio_vec.offset as _);
+                buf_len += bio_vec.len;
+            }
+        }
+
+        let region_bytes = region.len() * BLOCK_SIZE;
+        if region_bytes != buf_len {
```
Comment (lucassong-mh) [inline linux/dm-sworndisk/src/bio.rs:80]: Can we put length mismatch check in the front of this function?

File: linux/dm-sworndisk/src/block_device.rs [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    /// Returns the block range of the host `block_device`.
+    pub fn region(&self) -> Range<usize> {
+        // SAFETY: `self.inner` is valid `block_device`.
+        let start_sector = unsafe { (*self.inner.as_ptr()).bd_start_sect as usize };
+        let nr_sectors = unsafe { (*self.inner.as_ptr()).bd_nr_sectors as usize };
+        let end_sector = start_sector + nr_sectors;
+
+        Range {
+            start: (start_sector + BLOCK_SECTORS - 1) / BLOCK_SECTORS,
+            end: end_sector / BLOCK_SECTORS,
```
Comment (lucassong-mh) [inline linux/dm-sworndisk/src/block_device.rs:68]: The calculation logic of sector to block range deserve a comment.

File: linux/bindings/src/crypto/mod.rs [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+struct SgTable(Opaque<crate::sg_table>);
+
+impl SgTable {
+    /// Allocates a scatterlist table with `nents` entries.
+    pub fn alloc(nents: usize) -> Result<Self> {
+        let mut table = crate::sg_table::default();
+        // SAFETY: `table` is non-null and valid.
+        unsafe {
+            kernel::error::to_result(crate::sg_alloc_table(
+                core::ptr::addr_of_mut!(table),
```
[the remark this answers is not in the document]
Comment (cqs21) [reply linux/bindings/src/crypto/mod.rs:60]: Creating a reference with `&/&mut` is only allowed if the pointer is properly aligned and points to initialized data. And `table` will  actually be initialized by `sg_alloc_table`, so I thought `addr_of_mut!` is better.

File: linux/dm-sworndisk/src/block_device.rs [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+            .field("region", &self.region())
+            .finish()
+    }
+}
+
+impl Drop for HostBlockDevice {
+    fn drop(&mut self) {
+        // SAFETY: `self.inner` is a valid `block_device`, we get a reference
+        // to it when `open`, now put it here.
+        unsafe { bindings::blkdev_put(self.inner.as_ptr(), core::ptr::null_mut()) }
```
[the remark this answers is not in the document]
Comment (cqs21) [reply linux/dm-sworndisk/src/block_device.rs:85]: We will get a reference by `blkdev_get_by_path` when `open`.

File: linux/dm-sworndisk/src/bio.rs (hunk 6 above)
Comment (cqs21) [reply linux/dm-sworndisk/src/bio.rs:80]: Sure.

File: linux/bindings/src/crypto/skcipher.rs (hunk 3 above)
Comment (cqs21) [reply linux/bindings/src/crypto/skcipher.rs:138]: To avoid confusion when switching back and forth between the C and Rust sides.

File: linux/bindings/src/crypto/skcipher.rs (hunk 2 above)
Comment (cqs21) [reply linux/bindings/src/crypto/skcipher.rs:121]: That's a good point.

File: linux/bindings/src/crypto/aead.rs (hunk 1 above)
Comment (cqs21) [reply linux/bindings/src/crypto/aead.rs:267]: When `src` is dropped, `sg_free_table` will be called to free the memory.

File: linux/bindings/src/crypto/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (tatetian) [reply linux/bindings/src/crypto/mod.rs:29]: But in this context, which is the `drop` method, the pointer must be valid.

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!
